### ISSUE TITLE
FEATURE REQUEST Online benchmarking and ranking of published solutions

### ISSUE THREAD

**Issue by avegner** [exercism/exercism#5043]
Hi there!
The site doesn't allow to compare one's published solution with others in terms of time, mem and code size (e.g. in relevant symbols) stats.
It'd be very helpful for learners to have a way to compare their algorithms with others.
Please consider online benchmarking and ranking for your roadmap.
I had to create my own benchmarking and ranking tool for Go track especially:
https://github.com/avegner/exercism-bench

**Comment by iHiD** [comment]
Hello :)

> It'd be very helpful for learners to have a way to compare their algorithms with others.

I think this is covered by https://github.com/exercism/exercism/issues/4067, so I suggest we continue that discussion there.

> Please consider online benchmarking and ranking for your roadmap.

Exercism's aim is not to rank students or algorithms, its to help people learn the idioms of a language. For some languages (such as Go) benchmarking is very important, and covered in the tests. In other languages, it's not that important. So we won't be adding these sorts of leaderboards etc to the main site.

> I had to create my own benchmarking and ranking tool for Go track especially:

Very cool that you're building tooling. How are you getting the solutions? Through the API or by scraping the site?

**Comment by avegner** [comment]
> I think this is covered by #4067, so I suggest we continue that discussion there.

Ok, thanks for the reference.

> Exercism's aim is not to rank students or algorithms, its to help people learn the idioms of a language. For some languages (such as Go) benchmarking is very important, and covered in the tests. In other languages, it's not that important. So we won't be adding these sorts of leaderboards etc to the main site.

I get you. But I've seen a lot of suboptimal or even overkill solutions. Idioms are very important.
But a proper choice of tools (stdlib's functional, libs, data types, algorithms, tradeoffs) isn't less important since in the real world of programming knowing only idioms isn't enough.
How can one learn better? Only by comparing own solution with others. Not for leadboards but only for learning...

> Very cool that you're building tooling. How are you getting the solutions? Through the API or by scraping the site?

By scraping the site for now. If it violates any policies or rules (or whatever) I'll use the API (if it's officially open and may be accessed by unofficial tools).

**Comment by iHiD** [comment]
> Only by comparing own solution with others (better ones).

I agree this is super important. If you check out my final comment on the post I linked to, that discusses how we can group solutions etc. I'd love to then provide some teaching resources (blogs, videos, etc) around different approaches that people use etc. Super excited about this :)

So I agree about grouping/comparing - we're just going to have a "leaderboard" approach :)

> If it violates any policies or rules (or whatever) I'll use the API

I'm happy for you do this, but it would be friendly to us if you could add a `sleep(1second)` in between requests so you don't overload our servers or slow things down for others :)

**Comment by Luapulu** [comment]
Just came here to say that benchmarking like this would be really cool for the julia track, too. I'm happy for this to not be emphasised much on the site, but it'd be really useful to know if there are faster solutions out there and how you rank. Mentors rarely tell you if you've got the fastest solution / how to make it faster, so it'd be good to get some extra feedback.

**Comment by kotp** [comment]
> Mentors rarely tell you if you've got the fastest solution / how to make it faster, so it'd be good to get some extra feedback.

I wonder if that is because "students rarely ask".  The idea is probably idiomatic Julia, rather then "the fastest Julia".  But I am sure if one were to ask, the frequency of getting an answer to "how to do 'faster'" would increase.  It isn't generally the focus for the platform.

**Comment by ErikSchierboom** [comment]
Hello 👋

With the launch of Exercism v3, we are **closing all issues in this repository** to help give us a clean slate to detect new problems. If this issue is still relevant to Exercism v3 (e.g. it's a feature that we haven't implemented in v3, or a bug that still exists), please reopen it and we will review it and post an update on it as soon as we get chance.

Thanks for helping make Exercism better, and we hope you enjoy v3 🙂

**Comment by MeerKatDev** [comment]
I'd like to reopen this issue. I think timing the iterations would be a cool feature. Shouldn't take up much of the UI (the iteration bar has a lot of space in it) and it would be really useful for students and mentors alike. We would maybe keep a separated (relatively) cheap machine only for it, that runs the benchmarks asynchronously (that means, after submitting an iteration it would be added to a queue and appear some time later). 

I want to be clear though - I'm only requesting online benchmarking for single iterations, not ranking. I'd say there's HackerRank for that :)

**Comment by ErikSchierboom** [comment]
What do you mean exactly with benchmarking? Showing how long it takes for the test runner to run the tests for an iteration?

**Comment by MeerKatDev** [comment]
yes, exactly. Maybe also showing how long each test takes

**Comment by ErikSchierboom** [comment]
That could indeed be useful information. It would require all test runners to be updated though, so that's a lot more work.

**Comment by MeerKatDev** [comment]
Without updating the test runners, we could also get the timing at OS level, i.e. seeing how long it takes for a command to run from start to end. It may not be accurate, but it'd be a good baseline indicator I suppose. Test runners run on command line, so it'd be enough to wrap those commands, doesn't matter the language.

**Comment by ErikSchierboom** [comment]
I was referring to the "how long each test takes" :)

**Comment by MeerKatDev** [comment]
ooooo ok, yeah, that's definetely more work :)

**Comment by asarkar** [comment]
I’d also like to see some sort of “ranking” of solutions as I explained in https://github.com/exercism/exercism/issues/6395.
> Currently, there's a bunch of community solutions, some better than others, there's no way to tell how they "rank" against each other. If I want to learn from others, understandably, I only want to learn from the best. Same goes for a student's own solution, they may work hard to optimize the algorithm, or go for a naive one, it's all the same if it passes the tests.

It'll be nice to see each published solution rated as "80% faster (21ms), 33% less memory (2MB), out of 2951 total solutions". The number of total solutions at the time is important because it will obviously keep increasing over time.

Student should see the same data when they submit their own solution, published or not.
